src/Data/Graph.purs

```diff
@@ -4,32 +4,79 @@ module Data.Graph
   ( Graph
   , unfoldGraph
   , fromMap
+  , toMap
+  , empty
+  , insertEdge
+  , insertVertex
+  , insertEdgeWithVertices
   , vertices
   , lookup
   , outEdges
+  , children
+  , descendants
+  , parents
+  , ancestors
   , topologicalSort
+  , isInCycle
+  , isCyclic
+  , isAcyclic
+  , alterVertex
+  , alterEdges
+  , adjacent
+  , isAdjacent
+  , areConnected
+  , shortestPath
+  , allPaths
   ) where
 
 import Prelude
-import Data.Bifunctor (lmap)
+
+import Data.Array as Array
+import Data.Bifunctor (lmap, rmap)
 import Data.CatList (CatList)
 import Data.CatList as CL
 import Data.Foldable (class Foldable)
+import Data.Foldable as Foldable
 import Data.List (List(..))
 import Data.List as L
+import Data.List as List
 import Data.Map (Map)
 import Data.Map as M
-import Data.Maybe (Maybe(..), maybe)
-import Data.Tuple (Tuple(..), fst, snd)
+import Data.Maybe (Maybe(..), isJust, maybe)
+import Data.Set (Set)
+import Data.Set as S
+import Data.Set as Set
+import Data.Tuple (Tuple(..), fst, snd, uncurry)
 
 -- | A graph with vertices of type `v`.
 -- |
 -- | Edges refer to vertices using keys of type `k`.
-newtype Graph k v = Graph (Map k (Tuple v (List k)))
+newtype Graph k v = Graph (Map k (Tuple v (Set k)))
```

Comment:
This prevents the possibility of having multiple edges between any pair of nodes!